### PR TITLE
[1LP][RFR] Add support for create and nav_away to Host.update() and test_infrastructure_hosts_crud

### PR DESCRIPTION
**Pull request by prichard77**
## Purpose or Intent

- __Adding tests__ for host edit to support cancel of edit and also navigating away before saving updates. The navigating away case will cover customer BZ 1634794.

### PRT Run

{{ pytest: --use-provider complete --long-running cfme/tests/infrastructure/test_host.py::test_infrastructure_hosts_crud }}

### CODE REVIEW

**Comment by prichard77** [comment]
I am still making small updates to this and testing, but I'd like to start the review.

File: cfme/tests/infrastructure/test_host.py [hunk 1]
```diff
@@ -541,9 +541,10 @@ def test_add_ipmi_refresh(appliance, setup_provider):
 
 
 @test_requirements.infra_hosts
-@pytest.mark.parametrize("num_hosts", [1, 2, 4])
-@pytest.mark.meta(blockers=[BZ(1634794, forced_streams=["5.10"])], automates=[1634794])
-def test_infrastructure_hosts_crud(appliance, setup_provider_min_hosts, provider, num_hosts):
+@pytest.mark.meta(automates=[1634794])
+@pytest.mark.parametrize("crud_action", ['edit_from_hosts', 'edit_from_details', 'cancel',
+                                         'nav_away_changes', 'nav_away_no_changes', 'remove'])
```
Comment (john-dupuy) [inline cfme/tests/infrastructure/test_host.py:546]: In my opinion, and definitely just an opinion so feel free to disagree, the `crud` test should be only testing core functionality, i.e. can you create, update, and delete an object? The parametrization here is only around the `update` part, and therefore I think it should be in a separate test like `test_infrastructure_host_update_scenarios`. 

Then the crud test would only include the basic `update` scenario. What do you think?

File: cfme/tests/infrastructure/test_host.py [hunk 2]
```diff
@@ -553,16 +554,36 @@ def test_infrastructure_hosts_crud(appliance, setup_provider_min_hosts, provider
...
+        else:
+            try:
+                assert navigate_to(host, 'Details').entities.summary("Properties").get_text_of(
+                    "Custom Identifier") == existing_custom_id
+            except NameError:
+                if existing_custom_id:
+                    raise
+    else:
+        host.delete()
+        host.delete(cancel=False)
```
[the remark this answers is not in the document]
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:589]: Including testing of canceling the delete. The first delete is cancelled.

[the remark this answers is not in the document]
Comment (john-dupuy) [reply cfme/tests/infrastructure/test_host.py:589]: Ah okay, that makes since, maybe explicitly include `cancel=True` in the first call so that is clear?

File: cfme/common/host_views.py [hunk 3]
```diff
@@ -302,8 +301,13 @@ class HostEditView(HostFormView):
     breadcrumb = BreadCrumb()
     save_button = Button("Save")
     reset_button = Button("Reset")
+    cancel_button = Button("Cancel")
 
-    is_displayed = displayed_not_implemented
+    @property
+    def is_displayed(self):
+        return (
+            self.logged_in_as_current_user and self.navigation.currently_selected == [
```
[the remark this answers is not in the document]
Comment (prichard77) [reply cfme/common/host_views.py:309]: I am updating is_displayed to something more specific and removing self.logged_in_as_current_user. Thanks

File: cfme/tests/infrastructure/test_host.py (hunk 1 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:546]: I disagree. I don't really have knowledge of the history of the usage of CRUD within CFME, but I  am only testing that we edit and delete here like I would expect. there are a few variations to deal with the different places you can edit from. It seemed like a natural place to add these. If this is really something the team discourages, I could make a change, but I'd question why we are getting so particular with the naming of the test. Isn't coverage the end goal? What would splitting the tests up accomplish? I parameterize around edit and delete functions, not just edit.

File: cfme/tests/infrastructure/test_host.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:589]: Roger.

File: cfme/tests/infrastructure/test_host.py (hunk 1 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:546]: I could split this up, I just want to make sure this is what we want to do, It seems to me I am just making 2 tests out of one where I see much of the time we are trying to combine tests AFAICT.

File: cfme/tests/infrastructure/test_host.py [hunk 4]
```diff
@@ -553,16 +554,36 @@ def test_infrastructure_hosts_crud(appliance, setup_provider_min_hosts, provider
...
+                "Properties").get_text_of("Custom Identifier")
+        except NameError:
+            existing_custom_id = None
+    if crud_action != 'remove':
+        stamp = fauxfactory.gen_alphanumeric()
+        new_custom_id = f'Edit host data. {stamp}'
+        try:
+            with update(host, action=crud_action):
+                host.custom_ident = new_custom_id
+        except UnexpectedAlertPresentException as e:
```
Comment (jaryn) [inline cfme/tests/infrastructure/test_host.py:570]: I am puzzled. Do you mean to investigate the exception only if it happens? And if it happens only if the action is some specific one? This doesn't seem right.

If you expect the action to raise an exception, you should do something like this:
```
if crud_action in ['cancel', 'nav_away_no_changes']:
    with pytest.raises(UnexpectedAlertPresentException) match=r".*Abandon changes.*"):
        with update(host, action=crud_action):
            host.custom_ident = new_custom_id
```

The other catches are perhaps also candidates to be done like that.

Optional: I would think about a way to reduce the amount of branches. Perhaps something like I did in https://github.com/ManageIQ/integration_tests/pull/10019/files#diff-fc91ed43845421f3eda7a5bb17217dc7R279

File: cfme/tests/infrastructure/test_host.py [hunk 5]
```diff
@@ -553,16 +554,36 @@ def test_infrastructure_hosts_crud(appliance, setup_provider_min_hosts, provider
...
+    host = appliance.collections.hosts.all()[0]
+    if crud_action in ['cancel', 'nav_away_changes', 'nav_away_no_changes']:
+        try:
+            existing_custom_id = navigate_to(host, 'Details').entities.summary(
+                "Properties").get_text_of("Custom Identifier")
+        except NameError:
+            existing_custom_id = None
+    if crud_action != 'remove':
+        stamp = fauxfactory.gen_alphanumeric()
+        new_custom_id = f'Edit host data. {stamp}'
```
[the remark this answers is not in the document]
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:566]: Done.

**Comment by prichard77** [comment]
FYI, This PR includes commits from PR#9961. At the moment it contains all of the commits for PR#9961.

File: cfme/infrastructure/host.py [hunk 6]
```diff
@@ -92,15 +92,29 @@ def __init__(self, **kwargs):
...
 
-    def update(self, updates, validate_credentials=False):
+    def update(self, updates, validate_credentials=False, action='edit_from_details'):
         """Updates a host in the UI. Better to use utils.update.update context manager than call
         this directly.
 
         Args:
            updates (dict): fields that are changing.
+           action (str): denotes additional functionality. Expecting edit_from_details,
+           edit_from_hosts, delete, cancel, or nav_away.
```
Comment (mshriver) [inline cfme/infrastructure/host.py:102]: Why would you expect `delete` as an action within update? It doesn't look like its actually handled in the logic here.

Comment (mshriver) [inline cfme/infrastructure/host.py:102]: For `cancel`, we already have consistently used a specific boolean kwarg for this behavior within CRUD methods.

https://github.com/ManageIQ/integration_tests/search?q=%22cancel%3DFalse%22&unscoped_q=%22cancel%3DFalse%22

I don't think this update method should wrap that into the action - as its mutually exclusive with the navigation path used to get to update.

File: cfme/infrastructure/host.py [hunk 7]
```diff
@@ -92,15 +92,29 @@ def __init__(self, **kwargs):
...
             self.ipmi = kwargs.get('ipmi')
 
-    def update(self, updates, validate_credentials=False):
+    def update(self, updates, validate_credentials=False, action='edit_from_details'):
         """Updates a host in the UI. Better to use utils.update.update context manager than call
         this directly.
 
         Args:
            updates (dict): fields that are changing.
+           action (str): denotes additional functionality. Expecting edit_from_details,
```
Comment (mshriver) [inline cfme/infrastructure/host.py:101]: I see how you're trying to consolidate on the test function calls by pushing options into `action` here, but its inconsistent with the rest of the entity update methods, and I don't think its something we want to build into the framework in this way.

Just using bare strings is going to be an issue, especially if we try to apply this kwarg to other entity update methods.

That aside, nearly all entity update methods are written to navigate to the entity `Edit` destination (which uses `Details` as prerequisite), and then fill the form.  Test cases parametrize on the nav path, and navigate + form fill directly if not navigating from details.

I don't think that the `Updateable` mixin / `update` contextmanager will actually handle the `cancel` action here either - unless an exception is raised on `cancel=True`, the contextmanager will return an updated entity instance. The entity attributes will no longer match what's in MIQ.

https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/update.py#L54

File: cfme/tests/infrastructure/test_host.py [hunk 8]
```diff
@@ -553,16 +554,36 @@ def test_infrastructure_hosts_crud(appliance, setup_provider_min_hosts, provider
...
-    hosts_view.toolbar.configuration.item_select('Edit Selected items',
-                                                handle_alert=False)
-    try:
-        hosts_view.navigation.select('Compute', 'Infrastructure', 'Providers', handle_alert=False)
-        final_view = provider.create_view(InfraProvidersView)
-        assert final_view.is_displayed
-    except UnexpectedAlertPresentException:
-        pytest.fail("Abandon changes alert displayed, but no changes made.")
-    # TODO add additional crud functionality.
+    host = appliance.collections.hosts.all()[0]
```
[the remark this answers is not in the document]
Comment (mshriver) [reply cfme/tests/infrastructure/test_host.py:557]: @jaryn good pattern suggestion - but here I don't think we want an exception or test failure when there are multiple hosts, we just want to pick the first one.

File: cfme/tests/infrastructure/test_host.py (hunk 1 above)
Comment (mshriver) [reply cfme/tests/infrastructure/test_host.py:546]: @john-dupuy thanks for voicing the opinion, but I think the parametrization of the CRUD test function is a good way to consolidate the test coverage here, as long as the parametrization is focused on CRUD paths.

For example, I wouldn't want the CRUD parametrization to include the credentials update coverage, or any comparisons.

I think that including edit via details / all selection and delete via the same selection options is appropriate to include in CRUD parametrization, because they're just the different ways that the entity can be updated/deleted from MIQ. To me this equivalent to parametrization of the ViaUI/ViaREST contexts for CRUD.

File: cfme/infrastructure/host.py (hunk 6 above)
Comment (prichard77) [reply cfme/infrastructure/host.py:102]: You are correct. Removing

File: cfme/tests/infrastructure/test_host.py (hunk 4 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:569]: This is to catch an exception that was thrown before the fix for BZ 1634794 was merged. This has been fixed but I want to catch it if it occurs again.

File: cfme/infrastructure/host.py (hunk 7 above)
Comment (prichard77) [reply cfme/infrastructure/host.py:101]: I was pretty confident the cancel stuff was working. Maybe I missed something, but I tested this when I don't click save in update, the custom_ID is the same as it was before I ran the test. Anyway, the bigger issue is how do we want to implement all of the cases without actions? Do I add a bunch or options kwargs instead? Suggestions? Maybe we should focus back on PR#9761 and work on this after it's merged? Thoughts?

**Comment by prichard77** [comment]
I am setting this back to WIPTEST as I have a few things to work out based upon comments, and I want reviewers to focus on other PRs for the moment.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/infrastructure/host.py [hunk 9]
```diff
@@ -92,15 +92,29 @@ def __init__(self, **kwargs):
             super(Host.Credential, self).__init__(**kwargs)
             self.ipmi = kwargs.get('ipmi')
 
-    def update(self, updates, validate_credentials=False):
+    def update(self, updates, validate_credentials=False, action='edit_from_details'):
```
[the remark this answers is not in the document]
Comment (prichard77) [reply cfme/infrastructure/host.py:95]: I took a look at varmeth. We can discuss further about updating in a nother PR, but at first glance I am not sure varmeth is the correct solution. I might could be convinced otherwise, but the pieces I would use to build the variable methods are all similar with just specific pieces added or removed. Coding for each context would create a bunch of duplicate code. Seems like adding conditionals via kwargs (or another solution) might work better.  IMHO

File: cfme/infrastructure/host.py (hunk 6 above)
Comment (prichard77) [reply cfme/infrastructure/host.py]: We do have a cancel kwarg implemented for delete. I use this in the test. I don't see the same functionality for edit/update. Maybe I should implement it for update.

File: cfme/infrastructure/host.py (hunk 7 above)
Comment (prichard77) [reply cfme/infrastructure/host.py:101]: I am going to implement actions as a few kwargs. If there is another suggested solution please comment.

**Comment by prichard77** [comment]
I am updating some comments in host.update that I forget to update, but otherwise changed usage of action string to kwargs. I'll be pushing changes to comments shortly, but would still like review comments.

File: cfme/infrastructure/host.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (john-dupuy) [reply cfme/infrastructure/host.py]: Yup, just a suggestion. It seems like it'd be useful for things like this, but I haven't tried it myself yet :smile:

File: cfme/tests/infrastructure/test_host.py [hunk 10]
```diff
@@ -540,26 +541,51 @@ def test_add_ipmi_refresh(appliance, setup_provider):
 
 
 @test_requirements.infra_hosts
-@pytest.mark.parametrize("crud_action", ['edit_from_hosts', 'edit_from_details', 'remove'])
+@pytest.mark.meta(automates=[1634794])
+@pytest.mark.parametrize("crud_action", ['edit_from_hosts', 'edit_from_details', 'cancel',
+                                         'nav_away_changes', 'nav_away_no_changes', 'remove'])
 def test_infrastructure_hosts_crud(appliance, setup_provider, crud_action):
     """
```
Comment (tpapaioa) [inline cfme/tests/infrastructure/test_host.py:548]: Add a description of what the test is doing for the different values of `crud_action`.

File: cfme/tests/infrastructure/test_host.py [hunk 11]
```diff
@@ -540,26 +541,51 @@ def test_add_ipmi_refresh(appliance, setup_provider):
...
     Polarion:
         assignee: prichard
         casecomponent: Infra
         caseimportance: low
         initialEstimate: 1/6h
+    Bugzilla:
+        1634794
     """
     host = appliance.collections.hosts.all()[0]
+    if crud_action in ['cancel', 'nav_away_changes', 'nav_away_no_changes']:
```
Comment (tpapaioa) [inline cfme/tests/infrastructure/test_host.py:558]: The if/else logic is hard to follow:

```
    if crud_action in ['cancel', 'nav_away_changes', 'nav_away_no_changes']:

    if crud_action != 'remove':

        if crud_action not in ['cancel', 'nav_away_changes', 'nav_away_no_changes']:

        else:

    else:
```

I think something like the following would be clearer:

```
if crud_action == 'remove':
    [delete() calls here]

elif crud_action in ('edit_from_hosts', 'edit_from_details'):
    [update() call here]

else:
    [navigate_away/cancel helper method call here]
```

File: cfme/tests/infrastructure/test_host.py [hunk 12]
```diff
@@ -540,26 +541,51 @@ def test_add_ipmi_refresh(appliance, setup_provider):
...
-        assert host.custom_ident == new_custom_id
-        assert navigate_to(host, 'Details').entities.summary("Properties").get_text_of(
-            "Custom Identifier") == new_custom_id
+        new_custom_id = f'Edit host data. {fauxfactory.gen_alphanumeric()}'
+        try:
+            with update(host,
+                        from_details=(crud_action == 'edit_from_details'),
+                        cancel=(crud_action == 'cancel'),
+                        nav_away=('nav_away' in crud_action),
+                        changes=('changes' in crud_action and 'no' not in crud_action)
```
Comment (tpapaioa) [inline cfme/tests/infrastructure/test_host.py:571]: Whether you keep these new kwargs in `update()` or move them elsewhere, it's better to be explicit about what values of `crud_action` correspond to which kwarg values:

```
'nav_away': crud_action in ('nav_away_changes', 'nav_away_no_changes'),
'changes': crud_action == 'nav_away_changes'
```

File: cfme/tests/infrastructure/test_host.py (hunk 4 above)
Comment (tpapaioa) [reply cfme/tests/infrastructure/test_host.py:574]: What does catching and re-raising exceptions accomplish here and in L583-L588? If `update()` raises an exception, then pytest logs it and the test fails, regardless.

File: cfme/infrastructure/host.py (hunk 7 above)
Comment (prichard77) [reply cfme/infrastructure/host.py:102]: I had originally written this as it's own test. This is a customer BZ. I am beginning to have many one of tests for customer BZs and was trying to consolidate into another test that made sense logically. Can I get a little more on why exactly update is the wrong place for this? I understand I my initial implementation with strings is not ideal (I have already updated to use kwargs) but why not in update? If I put this functionality somewhere else, it's going to be a lot of duplication of host.update.

File: cfme/tests/infrastructure/test_host.py (hunk 4 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:574]: I was just trying to catch a specific exception that was caused by an old BZ. Maybe I should have mentioned the old BZ in the logging test. Basically, I just wanted it to be easier to tell that the BZ this test is created to catch is reoccurring when/if it does.  If this is something we really discourage, I can remove it but I think it makes the failure clearer.

Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:574]: I added BZ number to log message. More input on this?

File: cfme/tests/infrastructure/test_host.py [hunk 13]
```diff
@@ -553,16 +554,36 @@ def test_infrastructure_hosts_crud(appliance, setup_provider_min_hosts, provider
...
+        try:
+            with update(host, action=crud_action):
+                host.custom_ident = new_custom_id
+        except UnexpectedAlertPresentException as e:
+            if crud_action in ['cancel', 'nav_away_no_changes'] and "Abandon changes" in e.msg:
+                pytest.fail("Abandon changes alert displayed, but no changes made.")
+            else:
+                raise
+        if crud_action not in ['cancel', 'nav_away_changes', 'nav_away_no_changes']:
+            assert host.custom_ident == new_custom_id  # is this actually doing anything here
```
[the remark this answers is not in the document]
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py]: It's a UI item. It can be seen on the details page and edit pages.

File: cfme/tests/infrastructure/test_host.py (hunk 10 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:548]: Added. Let me know if it's clear.

**Comment by dajoRH** [comment]
Lint report for commit 463973aa932cc7cc53e6a5b3dc5acdf04a304ebc:

`cfme/tests/infrastructure/test_host.py`:
- :red_circle: Line 3:1: **F401** *'datetime.datetime' imported but unused*
- :red_circle: Line 12:1: **F401** *'cfme.common.host_views.HostDetailsView' imported but unused*
- :red_circle: Line 13:1: **F401** *'cfme.common.host_views.HostEditView' imported but unused*
- :red_circle: Line 14:1: **F401** *'cfme.common.host_views.HostsCompareView' imported but unused*
- :red_circle: Line 17:1: **F401** *'cfme.common.provider_views.InfraProvidersView' imported but unused*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/infrastructure/test_host.py (hunk 11 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py]: I added comments and updated the structure of test_infrastructure_hosts_crud to make things clearer. Still adding comments where I think they would help clarify.

Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py]: I added comments and rearranged the structure of test_infrastructure_hosts_crud to make things clearer. I am still adding commants to host.update

File: cfme/tests/infrastructure/test_host.py (hunk 12 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:572]: This does make things clearer. Updated as suggested.

**Comment by prichard77** [comment]
I updated so that edit from details would be the default behavior as it was originally.

**Comment by prichard77** [comment]
rerunning PRT. I am not sure why it's failing as it is passing when I run locally.

File: cfme/tests/infrastructure/test_host.py (hunk 4 above)
Comment (mshriver) [reply cfme/tests/infrastructure/test_host.py:575]: I think its a good thing to catch this and fail explicitly with a prior BZ

File: cfme/infrastructure/host.py (hunk 7 above)
Comment (mshriver) [reply cfme/infrastructure/host.py]: I'd have to agree with @tpapaioa here, I don't think that the update method is the appropriate place for these more complex interaction paths.

It creates an inconsistent API for handling update on the entity, compared to all the other Entity update methods.  These methods are generally just implemented to handle the standard positive path actions for editing an entity through the UI - where `cancel` kwarg is just controlling whether `Save` or `Cancel` is clicked after filling the form.  Many of these entity methods also include a `from_details` kwarg, that controls how the automation gets to the edit page.  That should not be changed here to `from_hosts` as again that makes for an inconsistent API for the entity CRUD methods.

I don't think this is a pattern of behavior that we should push to all the other Entity classes with update methods, so I think its better to capture this specific behavior within the test as Tasos suggested.

There will be some duplication of what's in host.update within the test, but not that much, and it will be specific to the UI behavior that we're trying to explicitly test.  The specific navigation cases that you're trying to test don't require you to duplicate all of the behavior in update, just what's relevant to the test. We're not trying to capture a large permutation matrix of the actions, just a few specific paths.  This may make the test code a bit longer, but the behavior is specific to this test coverage - not behavior that we'll want to regularly replicate on a host for other test coverage.

To summarize, please:
1. Keep `from_details` kwarg on `Host.update` method
2. Add `cancel` kwarg to `Host.update`, given that it works properly in the context manager
3. Update the test to take the navigation steps for the actions that you want to test

In updating the test - the action paths don't have to be parametrized, as several can be covered in one test case flow. You can update with cancel, update from details, update from hosts, update and nav away with no changes, update and nav away with changes, and then delete, all in one flow.

Comment (prichard77) [reply cfme/infrastructure/host.py]: Updated as suggested.

File: cfme/infrastructure/host.py [hunk 14]
```diff
@@ -94,21 +94,25 @@ def __init__(self, **kwargs):
...
+            updates (dict): fields that are changing.
+            validate_credentials (bool): if True, validate host credentials
+            from_details (bool): if True, select 'Edit Selected items' from details view
+                else, select 'Edit Selected items' from hosts view
+            cancel (bool): click cancel button to cancel the edit if True
         """
-        if from_details:
-            view = navigate_to(self, "Edit")
-        else:
+        if not from_details:
```
Comment (mshriver) [inline cfme/infrastructure/host.py:108]: Can you leave the original logic in place here?  I don't see a benefit to switching these, or making this logic negative.

File: cfme/infrastructure/host.py [hunk 15]
```diff
@@ -136,7 +140,8 @@ def update(self, updates, validate_credentials=False, from_details=True):
             if validate_credentials:
                 view.endpoints.ipmi.validate_button.click()
         view.flash.assert_no_error()
-        changed = any([changed, credentials_changed, ipmi_credentials_changed])
+        changed = False if cancel else any([changed, credentials_changed,
```
Comment (mshriver) [inline cfme/infrastructure/host.py:143]: I think this buries the intent a bit here, and it would be better to adjust L145 to be:
```python
if changed and not cancel:
```

File: cfme/tests/infrastructure/test_host.py [hunk 16]
```diff
@@ -526,26 +530,100 @@ def test_add_ipmi_refresh(appliance, setup_provider):
...
-        stamp = fauxfactory.gen_alphanumeric()
-        new_custom_id = f'Edit host data. {stamp}'
 
-        with update(host, from_details=(crud_action == 'edit_from_details')):
-            host.custom_ident = new_custom_id
+    # Case1 - edit from Hosts
+    new_custom_id = f'Edit host data. {fauxfactory.gen_alphanumeric()}'
+    with update(host,
+                from_details=False,
+                cancel=False,
```
Comment (mshriver) [inline cfme/tests/infrastructure/test_host.py:550]: This test flow looks a lot better IMO! Thanks for making the switch.

Consider not passing cancel kwarg here, as `False` is the default. Optional change, being explicit is good too.

Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:550]: Yeah. I knew it wasn't required, but I thought it made the test clearer. I could go either way.

File: cfme/infrastructure/host.py (hunk 14 above)
Comment (prichard77) [reply cfme/infrastructure/host.py:108]: I agree. I was trying to keep the logic clearer, but I may have been a bit lazy when I changed the kwarg back to from_details. Updating.

File: cfme/infrastructure/host.py (hunk 15 above)
Comment (prichard77) [reply cfme/infrastructure/host.py:143]: Would be clearer, thanks.

File: cfme/tests/infrastructure/test_host.py (hunk 16 above)
Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:550]: I am going to remove the instances of cancel=False. I believe the logic can be understood without them. Thanks.

Comment (prichard77) [reply cfme/tests/infrastructure/test_host.py:550]: I went ahead and removed the cancel=False also from host.delete. Especially with the comment in the test, the host.delete logic should be clear.
